pkg/sink/sink.go

```diff
@@ -136,3 +167,96 @@ func createRequestURI(apiVersion, namePlural, namespace string) string {
 	uri = path.Join(uri, namePlural)
 	return uri
 }
+
+func (r Resource) validateEvent(triggerValidate *triggersv1.TriggerValidate, headers http.Header, payload []byte) error {
+	tr, err := r.createValidateTask(triggerValidate, headers, payload)
+	if err != nil {
+		return nil
+	}
+
+	tr, err = r.PipelineClient.TektonV1alpha1().TaskRuns(r.EventListenerNamespace).Create(tr)
+	if err != nil {
+		return err
+	}
+
+	for i := taskRunPollingRetries; i >= 0; i-- {
+		tr, err := r.PipelineClient.TektonV1alpha1().TaskRuns(r.EventListenerNamespace).Get(tr.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		if tr.IsSuccessful() {
+			break
+		}
+
+		if tr.IsDone() && !tr.IsSuccessful() {
+			return errors.New("validation taskrun: " + tr.Name + " failed")
+		}
+
+		time.Sleep(taskRunPollingInterval)
+	}
+	return nil
+}
+
+func (r Resource) createValidateTask(triggerValidate *triggersv1.TriggerValidate,
+	headers http.Header, payload []byte) (*pipelinev1.TaskRun, error) {
+	// Checking whether task define in taskref exists or not
+	task, err := r.PipelineClient.TektonV1alpha1().Tasks(r.EventListenerNamespace).Get(triggerValidate.TaskRef.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	if task.Spec.Inputs == nil {
+		return nil, errors.New("parameters are mandatory for validate task")
+	}
+
+	params := []pipelinev1.Param{}
+	params = append(params, triggerValidate.Params...)
+	params = append(params, pipelinev1.Param{
+		Name: "Payload",
+		Value: pipelinev1.ArrayOrString{
+			Type:      pipelinev1.ParamTypeString,
+			StringVal: string(payload),
+		},
+	})
+
+	for _, v := range task.Spec.Inputs.Params {
+		h, ok := headers[v.Name]
+		if !ok {
+			continue
+		}
+		if v.Type == pipelinev1.ParamTypeArray {
+			params = append(params, pipelinev1.Param{
+				Name: v.Name,
+				Value: pipelinev1.ArrayOrString{
+					Type:     v.Type,
+					ArrayVal: h,
+				},
+			})
+		} else if v.Type == pipelinev1.ParamTypeString {
+			params = append(params, pipelinev1.Param{
+				Name: v.Name,
+				Value: pipelinev1.ArrayOrString{
+					Type:      v.Type,
+					StringVal: h[0],
```

Comment:
Since the header key can be specified [multiple times](https://stackoverflow.com/questions/3096888/standard-for-adding-multiple-values-of-a-single-http-header-to-a-request-or-resp), maybe we should enforce the header to be passed an array always. Or at least conditionally check if there are multiple values?

Reply:
I like the array idea though it will restrict where the header values can be used. Taskruns have validation to ensure arrays can only be used in certain fields so this might make the header params less usable: https://github.com/tektoncd/pipeline/blob/7a0454b27e9966ac6063b4c5320c46864fe33fc6/pkg/apis/pipeline/v1alpha1/task_validation.go#L205

Reply:
Joining the array on space or some other delimiter was what I meant behind flattening. If we are limited by arrays, maybe this is preferable?